lib/spack/spack/database.py

```diff
@@ -339,7 +365,141 @@ def read_transaction(self):
+        err = 'Unable to mark {0.name} as failed.'
+
+        prefix = spec.prefix
+        if prefix not in self._prefix_failures:
+            mark = Lock(
+                self.prefix_fail_path,
+                start=spec.dag_hash_bit_prefix(bit_length(sys.maxsize)),
+                length=1,
+                default_timeout=self.package_lock_timeout, desc=spec.name)
+
+            try:
+                mark.acquire_write()
+                tty.debug('PID {0} succeeded in marking failure for {1}'
+                          .format(os.getpid(), spec.name))
+            except LockTimeoutError:
+                # Unlikely that another process failed to install at the same
+                # time but log it anyway.
+                tty.debug('PID {0} failed to mark install failure for {1}'
+                          .format(os.getpid(), spec.name))
+                tty.warn(err.format(spec))
+
+            # Whether we or another process marked it as a failure, track it
+            # as such locally.
+            self._prefix_failures[prefix] = mark
+
+        return self._prefix_failures[prefix]
+
+    def prefix_failed(self, spec):
+        """Return True if the prefix (installation) is marked as failed."""
+        # The failure was detected in this process.
+        if spec.prefix in self._prefix_failures:
+            return True
+
+        # The failure was detected by a concurrent process (e.g., an srun),
+        # which is expected to be holding a write lock if that is the case.
+        if self.prefix_failure_locked(spec):
+            return True
+
+        # Determine if the spec may have been marked as failed by a separate
+        # spack build process running concurrently.
+        return self.prefix_failure_marked(spec)
+
+    def prefix_failure_locked(self, spec):
+        """Return True if a process has a failure lock on the spec."""
+        try:
+            check = Lock(
+                self.prefix_fail_path,
+                start=spec.dag_hash_bit_prefix(bit_length(sys.maxsize)),
+                length=1,
+                default_timeout=self.package_lock_timeout, desc=spec.name)
+
+            check.acquire_read()
+
+            # If we have a read lock then no other process has a failure lock
+            # indicating an active installation failure for the spec.  There
+            # is no reason to hang on to the read lock itself.
+            check.release_read()
+            del check
+        except LockTimeoutError:
+            # Installation of the prefix has failed in another process holding
+            # a write lock.
+            tty.debug('{0} is failure locked'.format(spec.name))
+            return True
+
+    def prefix_failure_marked(self, spec):
+        """Determine if the spec has a persistent failure marking."""
+        try:
+            path = self._failed_spec_path(spec)
+            with open(path) as f:
+                spec2 = spack.spec.Spec.from_json(f)
```

Comment:
```suggestion
                spec_from_json = spack.spec.Spec.from_json(f)
```

Reply:
Excellent point.  Changed `spec2` to `spec_from_json`.